[10.0] [ADD] migrate packaging management

lmignon opened the pull request:
#61 forward ported in 10.0

Commit:
add modules to use product packaging

packaging_extended: to use uom in packaging
purchase_packaging: to improve supplierinfo with purchase uom and packaging and use them in purchase order
sale_packaging: to use packaging uom in sale order line

Commit:
purchase_packaging : po from procurement

When Purchase Order are created from procurement take account of seller configuration.

Commit:
rename packaging_extended to packaging_uom

Commit:
create purchase order from view

Commit:
purchase_packaging : compute price_unit with the new uom / packaging_management : update readme

Commit:
When create po line set product_qty to avoid to reset qty to 1

Commit:
remove useless xml view

Commit:
correct product creation with packaging

Commit:
correct xml view

Commit:
Improve create of purchase order line to set it compatible with test_data

Commit:
add missing decorators

Commit:
purchase_packaging : fix supplier info view for create

Commit:
purchase_packaging : add post_init_hook to correctly initialize product_purchase_qty

lmignon committed:
[MIG] packaging_uom, purchase_packaging, sale_packaging: Migrated to 10.0

rousseldenis commented on packaging_uom/models/product_packaging.py:
```diff
@@ -0,0 +1,45 @@
+        Compute the quantity by package based on uom
+        """
+        if self.uom_id and self.product_tmpl_id:
+            self.qty = self.uom_id. _compute_quantity(
```
I think there is a unwanted space after the dot

rousseldenis commented on sale_packaging/models/sale_order_line.py:
```diff
@@ -0,0 +1,38 @@
+        return vals
+
+    @api.model
+    @api.returns('self', lambda rec: rec.id)
```
Not needed anymore as this was for new/old api translation

lmignon replied:
@rousseldenis This decorator is still present into Odoo https://github.com/odoo/odoo/blob/10.0/odoo/models.py#L3772

rousseldenis replied:
Ok, that's because I referred to https://www.odoo.com/documentation/10.0/reference/orm.html => returns() description

lmignon committed:
[IMP] packaging_uom: declare 'inverse' method on qty
This change is required to make the code working with existing addons without breaking tets. By declaring an invers on qty, we are able to assign/create the required uom to the package based on the expected qty

rousseldenis committed:
[IMP] Adds the glue which fill in the product packaging in stock

When a sale order is confirmed and product packaging is filled in
sale order line(s), the information is propagated through
procurement orders and stock moves.
This is conditioned by a parameter on procurement rule.

Cedric-Pigeon commented on packaging_uom/models/product_packaging.py:
```diff
@@ -0,0 +1,82 @@
+            ("factor", "=", 1.0 / self.qty),
+            ('category_id', '=', category_id.id)])
+        if not uom_id:
+            uom_id = self.uom_id    .create({
```
Remove spaces

Cedric-Pigeon commented on purchase_packaging/models/procurement_order.py:
```diff
@@ -0,0 +1,30 @@
+        seller = self.product_id._select_seller(
+            partner_id=supplier.name,
+            quantity=res['product_qty'],
+            date=po.date_order and po.date_order[:10],
```
It should be better to use fields.date method to format it. no ?

Cedric-Pigeon commented on purchase_packaging/models/purchase_order_line.py:
```diff
@@ -0,0 +1,179 @@
+        return vals
+
+    @api.model
+    @api.returns('self', lambda rec: rec.id)
```
is it still usefull ?

Cedric-Pigeon reviewed (commented):
Minor comments

MiquelRForgeFlow commented on purchase_packaging/models/purchase_order_line.py:
```diff
@@ -0,0 +1,179 @@
+                ('uom_type', '=', 'reference')], limit=1)
+            to_uom = to_uom and to_uom[0]
+            self.product_qty = supplier.min_qty_uom_id._compute_quantity(
+                supplier.min_qty, to_uom
```
You need to convert to the uom of the supplier, if one was found in https://github.com/OCA/stock-logistics-warehouse/pull/259/files#diff-0b094619c99f4fabbe358caff73c8626R119. 

For example, if you have a minimum of 1 Half Dozen's of a product with packaging of Dozen's. When you order this product you will get by default 1 Half Dozen as Purchase UoM and 1 Dozen as Uom. You should get 1 Half Dozen and 0.5 Dozen.

rousseldenis replied:
@mreficent Purchase unit of measure in this case are a coefficent of the product unit of measure.

So, if your minimal purchase quantity is for example 0.5 x Dozen (Purchase UOM) of Dozen (Product UOM), the purchase order line will show in product_quantity 6 x Dozen (which is 0.5 x 12 x 12)

If your minimal quantity is 0.5 x Unit (Purchase UOM) of Dozen (Product UOM), the purhase order line will show in product quantity 0.5 x Dozen (which is 0.5 x 1 x 12).

If your product unit of measure is Unit, you can have a minimal quantity of 0.5 x Dozen (Purchase UOM) of Unit (Product UOM), that will be translated in 6 x Unit.

MiquelRForgeFlow commented:
It is hard to understand the supplier info screen. 
![image](https://user-images.githubusercontent.com/25005517/29770984-df8329d0-8bf1-11e7-8975-56cd8a354575.png)

MiquelRForgeFlow commented:
@lmignon Without fixing the above errors found it is difficult to continue the review. Can you look into it?

rousseldenis commented:
@mreficent I'm looking at this

rousseldenis commented:
![image](https://user-images.githubusercontent.com/19529533/29775000-d7d3de20-8c03-11e7-9569-67b0f044f62b.png)

@mreficent Is this more readable ?

rousseldenis commented:
Or this one (change the unit to understand)

![image](https://user-images.githubusercontent.com/19529533/29775128-4345693a-8c04-11e7-9ec7-7e71813b5b31.png)

MiquelRForgeFlow commented:
```
1.00 Dozen(s) of Unit(s)
of package COCA 1.5L LIGHT
```

rousseldenis commented:
@mreficent Didn't succeed in putting space between 1.00 and unit so I've added a x:

![image](https://user-images.githubusercontent.com/19529533/29777393-23065050-8c0c-11e7-8dc1-46f6b9df21d2.png)


If it's ok, I fix the other comments and try to add tests for codecoverage

JordiBForgeFlow commented:
@rousseldenis can you show "using package" instead of "of package"? There are now 2 "of"

rousseldenis commented:
@jbeficent We are open to change so ;-). Yes that's less confusing!

rousseldenis commented:
@Cedric-Pigeon @jbeficent @mreficent Fixed comments and other stuffs

rousseldenis commented:
@pedrobaeza

pedrobaeza commented:
Can this be squashed any way or should I merge this way?

rousseldenis commented:
@pedrobaeza Can you wait a little on this ?

We have an issue.

pedrobaeza commented:
Of course. Let me know

Commit:
[FIX] fix orderpoint compute with purchase multiple

Cedric-Pigeon commented:
I just added a failing test that shows up a bug.
In v8.0 it was working thanks to this line https://github.com/odoo/odoo/blob/8.0/addons/purchase/purchase.py#L1558

but in V10.0 Odoo changed its way of compute.

I am still looking for a fix. Any help is welcome.

rousseldenis commented on purchase_packaging/models/procurement_order.py:
```diff
@@ -28,3 +29,39 @@ def _prepare_purchase_order_line(self, po, supplier):
+            procs = self.env['procurement.order'].search(
+                [('orderpoint_id', '=', orderpoint.id),
+                 ('state', 'not in', ['cancel', 'done'])])
+            if procs:
```
```
po_lines = procs.mapped('purchae_line_id').filtered(lambda x: x.state == 'draft')
if po_lines:
    qty = ...
```

rousseldenis commented:
@pedrobaeza Problem solved. Commit history is relevant.
